### PR TITLE
fix(ivy): Add `style="{{exp}}"` based interpolation

### LINKED ISSUE #33575: Internal compiler error on style interpolation `style="{{exp}}"` with Ivy
# 🐞 bug report

### Affected Package
@angular/compiler

### Is this a regression?
Yes, works without ivy

### Description
Interpolating the style of a component with 

> `style = {{style}}`

Throws error in ivy

## 🔬 Minimal Reproduction
https://github.com/EmaGht/angularStyleInterpolation

instructions in README

## 🌍  Your Environment

**Angular Version:**
9.0.0-rc0

**Anything else relevant?**
Could you also, please, give a couple informations more when something like this happens?

When ivy crashes during template compilation it doesn't give any pointer to which file is actually affected by the error. It took me 5 hours of commenting out components to find this bug. 

Next time this happens, my sanity is on your conscience! 😭

### LINKED ISSUE #34056: [Ivy] [i18n] Translated messages are not loaded for library components
# 🐞 bug report

### Is this a regression?
Yes, the translated messages are picked up by Angular 6-8 and in Angular 9 with Ivy disabled.

### Description
Messages for third-party components are extracted successfully when running `ng xi18n`, but the compiler fails to locate the translated strings during compilation.

Possibly related: https://github.com/angular/angular/issues/32311, https://github.com/angular/angular/pull/32881

## 🔬 Minimal Reproduction
1. Clone https://github.com/tsvetomir/ivy-i18n-messages
1. Run `npm ci`
1. Run `ng serve --configuration=es` or `ng serve --configuration=es,production`

Result:
* None of the messages declared in `CalendarModule` are translated.
* The message declared in the application ("Hello, World!") is translated.
* If `enableIvy` is set to false in `tsconfig.json` messages are translated normally.

## 🔥 Exception or Error
No run-time errors.

## 🌍  Your Environment

**Angular Version:**
<pre><code>
Angular CLI: 9.0.0-rc.3
Node: 12.13.1
OS: darwin x64
Angular: 9.0.0-rc.3
... animations, cli, common, compiler, compiler-cli, core, forms
... language-service, platform-browser, platform-browser-dynamic
... router

Package                           Version
-----------------------------------------------------------
@angular-devkit/architect         0.900.0-rc.3
@angular-devkit/build-angular     0.900.0-rc.3
@angular-devkit/build-optimizer   0.900.0-rc.3
@angular-devkit/build-webpack     0.900.0-rc.3
@angular-devkit/core              9.0.0-rc.3
@angular-devkit/schematics        9.0.0-rc.3
@angular/localize                 9.0.0-rc.0
@ngtools/webpack                  9.0.0-rc.3
@schematics/angular               9.0.0-rc.3
@schematics/update                0.900.0-rc.3
rxjs                              6.5.3
typescript                        3.6.4
webpack                           4.41.2
</code></pre>

### PR DESCRIPTION
**Pull request by mhevery**
Fixes #33575

Add support for interpolation in styles as shown:
```
<div style="color: {{exp1}}; width: {{exp2}};">
```

## PR Checklist
Please check if your PR fulfills the following requirements:

- [ ] The commit message follows our guidelines: https://github.com/angular/angular/blob/master/CONTRIBUTING.md#commit
- [ ] Tests for the changes have been added (for bug fixes / features)
- [ ] Docs have been added / updated (for bug fixes / features)

## PR Type
What kind of change does this PR introduce?

- [ ] Bugfix
- [ ] Feature
- [ ] Code style update (formatting, local variables)
- [ ] Refactoring (no functional changes, no api changes)
- [ ] Build related changes
- [ ] CI related changes
- [ ] Documentation content changes
- [ ] angular.io application / infrastructure changes
- [ ] Other... Please describe:

## What is the current behavior?

Issue Number: N/A

## Does this PR introduce a breaking change?

- [ ] Yes
- [ ] No

### CODE REVIEW

File: packages/core/test/acceptance/styling_spec.ts [hunk 1]
```diff
@@ -303,6 +303,70 @@ describe('styling', () => {
...
+        <div style="content: &quot;{{one}}b{{two}}c{{three}}d{{four}}e&quot;"></div>
+        <div style="content: &quot;{{one}}b{{two}}c{{three}}d&quot;"></div>
+        <div style="content: &quot;{{one}}b{{two}}c&quot;"></div>
+        <div style="content: &quot;{{one}}b&quot;"></div>
+        <div style="{{self}}"></div>
+      `
+    })
+    class Cmp {
+      self = 'content: "self"';
+      one = 'one';
```
Comment (pkozlowski-opensource) [inline packages/core/test/acceptance/styling_spec.ts]: super nit: using actual number (ex. `one = 1`) would make it easier to "visually parse" asserts (`a1b2c ` vs. `aonebtwoc`

Comment (mhevery) [reply packages/core/test/acceptance/styling_spec.ts]: good idea, changed.

File: packages/compiler-cli/test/compliance/r3_view_compiler_styling_spec.ts [hunk 2]
```diff
@@ -1162,6 +1162,155 @@ describe('compiler compliance: styling', () => {
...
+    it('should support style interpolation', () => {
+      const files = {
+        app: {
+          'spec.ts': `
+                  import {Component, NgModule, HostBinding} from '@angular/core';
+  
+                  @Component({
+                    selector: 'my-component',
+                    template: \`
+                      <div style="p1:{{p1}};"></div>
```
Comment (sod) [inline packages/compiler-cli/test/compliance/r3_view_compiler_styling_spec.ts]: Does this reintroduce the trap as described in https://github.com/angular/angular/issues/8009, or does this write the properties individually into the style object?

Comment (sod) [reply packages/compiler-cli/test/compliance/r3_view_compiler_styling_spec.ts]: Hm, from https://hackmd.io/@5zDGNGArSxiHhgvxRGrg-g/rycZk3N5S it reads like it uses `style.cssText`, that works cross browser, nice 👍

Comment (mhevery) [reply packages/compiler-cli/test/compliance/r3_view_compiler_styling_spec.ts]: this should continue to work.

**Comment by mhevery** [comment]
[presubmit](http://test/OCL:292210551:BASE:292210599:1580334539457:869c6c1)

**Comment by surenrao** [comment]
I was getting `[error] Error: Internal Error: Unexpected interpolation` 
So in order to fix the issue, i searched `="{{` and replaced it with `[attr]=""`. Luckily only had 20-30 lines of codes change.

**Comment by mhevery** [comment]
[presubmit](http://test/OCL:292210551:BASE:294504501:1581455810343:7033334a)

File: packages/core/src/render3/instructions/style_map_interpolation.ts [hunk 3]
```diff
@@ -0,0 +1,336 @@
...
+ */
+
+import {getLView} from '../state';
+import {interpolation1, interpolation2, interpolation3, interpolation4, interpolation5, interpolation6, interpolation7, interpolation8, interpolationV} from './interpolation';
+import {ɵɵstyleMap} from './styling';
+
+
+/**
+ *
+ * Update an interpolated class on an element with single bound value surrounded by text.
```
Comment (pkozlowski-opensource) [inline packages/core/src/render3/instructions/style_map_interpolation.ts]: interpolated class -> interpolated style?

File: packages/core/src/render3/instructions/style_map_interpolation.ts [hunk 4]
```diff
@@ -0,0 +1,336 @@
...
+ *
+ * Used when the value passed to a property has 1 interpolated value in it:
+ *
+ * ```html
+ * <div class="prefix{{v0}}suffix"></div>
```
Comment (pkozlowski-opensource) [inline packages/core/src/render3/instructions/style_map_interpolation.ts]: class="prefix{{v0}}suffix" -> style="prefix{{v0}}suffix" ?

Here and in the entire file.

**Review by pkozlowski-opensource — APPROVED** [review]
LGTM for the runtime changes with some cleanups:
* see comments in-line;
* the commit message references #34056 as the issue being fixed here, but I'm not sure it is the right one?

File: packages/core/src/render3/instructions/style_map_interpolation.ts (hunk 4 above)
Comment (mhevery) [reply packages/core/src/render3/instructions/style_map_interpolation.ts]: fixed

**Comment by mhevery** [comment]
> * the commit message references #34056 as the issue being fixed here, but I'm not sure it is the right one?

You are right. Correct one is #33575

File: packages/core/src/render3/instructions/style_map_interpolation.ts [hunk 5]
```diff
@@ -0,0 +1,343 @@
...
+    i3: string, v4: any, i4: string, v5: any, i5: string, v6: any, i6: string, v7: any,
+    suffix: string): void {
+  const lView = getLView();
+  const interpolatedValue = interpolation8(
+      lView, prefix, v0, i0, v1, i1, v2, i2, v3, i3, v4, i4, v5, i5, v6, i6, v7, suffix);
+  ɵɵstyleMap(interpolatedValue);
+}
+
+/**
+ * Update an interpolated style on an element with 8 or more bound values surrounded by text.
```
Comment (kara) [inline packages/core/src/render3/instructions/style_map_interpolation.ts]: Shouldn't this be 9 or more? Interpolation8 handles 8.

```suggestion
 * Update an interpolated style on an element with 9 or more bound values surrounded by text.
```

File: packages/core/src/render3/instructions/style_map_interpolation.ts [hunk 6]
```diff
@@ -0,0 +1,343 @@
...
+ *
+ * Used when the number of interpolated values exceeds 7.
```
Comment (kara) [inline packages/core/src/render3/instructions/style_map_interpolation.ts]:
```suggestion
 * Used when the number of interpolated values exceeds 8.
```

File: packages/core/src/render3/instructions/style_map_interpolation.ts [hunk 7]
```diff
@@ -0,0 +1,343 @@
...
+ *  class="key: {{v0}}; key1: {{v1}}; key2: {{v2}}; key3: {{v3}}; key4: {{v4}}; key5: {{v5}};
+ *         key6: {{v6}}; key7: {{v7}}; key8: {{v8}}; key9: {{v9}}suffix"></div>
+ * ```
+ *
+ * Its compiled representation is:
+ *
+ * ```ts
+ * ɵɵstyleMapInterpolateV(
+ *    ['key: ', v0, '; key1: ', v1, '; key2: ', v2, '; key3: ', v3, '; key4: ', v4, '; key5: ', v5,
+ *     '; key6: ', v6, '; key7: ', v7, '; key9: ', v9, 'suffix']);
```
Comment (kara) [inline packages/core/src/render3/instructions/style_map_interpolation.ts]:
```suggestion
 *     '; key6: ', v6, '; key7: ', v7, '; key8', v8, '; key9: ', v9, 'suffix']);
```

File: packages/core/src/render3/instructions/style_map_interpolation.ts [hunk 8]
```diff
@@ -0,0 +1,343 @@
...
+ * ```
+ *.
+ * @param values The a collection of values and the strings in-between those values, beginning with
```
Comment (kara) [inline packages/core/src/render3/instructions/style_map_interpolation.ts]:
```suggestion
 * @param values The collection of values and the strings in-between those values, beginning with
```

**Review by kara — APPROVED** [review]
LGTM, with minor cleanup

**Comment by mhevery** [comment]
[presubmit](http://test/OCL:292210551:BASE:295059281:1581653368362:48e4a91f)

File: packages/core/src/render3/instructions/style_map_interpolation.ts (hunk 6 above)
Comment (mhevery) [reply packages/core/src/render3/instructions/style_map_interpolation.ts]: fixed

File: packages/core/src/render3/instructions/style_map_interpolation.ts (hunk 8 above)
Comment (mhevery) [reply packages/core/src/render3/instructions/style_map_interpolation.ts]: fixed

File: packages/core/src/render3/instructions/style_map_interpolation.ts (hunk 5 above)
Comment (mhevery) [reply packages/core/src/render3/instructions/style_map_interpolation.ts]: fixed

File: packages/core/src/render3/instructions/style_map_interpolation.ts (hunk 7 above)
Comment (mhevery) [reply packages/core/src/render3/instructions/style_map_interpolation.ts]: fixed

File: packages/core/test/acceptance/styling_spec.ts [hunk 9]
```diff
@@ -783,9 +846,10 @@ describe('styling', () => {
     expect(classList).not.toContain('two');
   });
 
-  it('should support interpolations inside a style property binding', () => {
-    @Component({
-      template: `
+  onlyInIvy('only Ivy supports style interpolation')
```
[the remark this answers is not in the document]
Comment (mhevery) [reply packages/core/test/acceptance/styling_spec.ts]: removed

**Comment by SkrupSulten** [comment]
Does anyone know when we can expect this fix to be out?

**Comment by mhevery** [comment]
Should be out later this week.

**Pull request closed without merging**